### PR TITLE
Py data source

### PR DESCRIPTION
**Pull request by wangkuiyi**
This PR depends on https://github.com/PaddlePaddle/Paddle/pull/1253.

Fixes https://github.com/PaddlePaddle/Paddle/issues/1264.

### CODE REVIEW

File: python/paddle/trainer_config_helpers/data_sources.py [hunk 1]
```diff
@@ -22,24 +22,22 @@
 except ImportError:
     import pickle
 
-__all__ = ['define_py_data_sources2']
+__all__ = ['setup_data_provider']
 
 
-def define_py_data_sources2(train_list, test_list, module, function, args=None):
+def setup_data_provider(train_list, test_list, module, function, args=None):
```
Comment (reyoung) [inline python/paddle/trainer_config_helpers/data_sources.py]: 为什么要改这个名字呢？

首先，对于使用Python驱动Paddle训练的API来说，不需要有DataProvider的概念了。而是会直接使用Python代码读取数据，然后传递给Paddle的C++库。这个函数不会被Python使用。

其次，修改名字会让非常多现在的模型的向后兼容成问题。

**Comment by reyoung** [comment]
为什么要改这个名字呢？

首先，对于使用Python驱动Paddle训练的API来说，不需要有DataProvider的概念了。而是会直接使用Python代码读取数据，然后传递给Paddle的C++库。这个函数不会被Python使用。

其次，修改名字会让非常多现在的模型的向后兼容成问题。

**Comment by wangkuiyi** [comment]
我的考虑是：如果要实现分布式训练，我们需要一个framework形式的API，也就是说main函数实现在C++代码里，main函数要能够执行用户写的Python程序来读取数据（因为用户数据格式千变万化，我们的C++程序恐怕不能预先都了解）。

以上是我在修改data provider的初衷。

不过要是我的理解有误（分布式训练也完全可以方便地通过API函数调用的方式，而不需要一个framework），我也很乐于看到简化（删除）data provider机制呀。

**Comment by reyoung** [comment]
> 我的考虑是：如果要实现分布式训练，我们需要一个framework形式的API，也就是说main函数实现在C++代码里，main函数要能够执行用户写的Python程序来读取数据（因为用户数据格式千变万化，我们的C++程序恐怕不能预先都了解）。

我理解这个意思了。

不过，我在想 训练用的main函数是写在C++里面的还是在Python里面的呢？ 网络通信和多机同步的策略肯定是在C++里面写，这不妨碍我们用一种机制将这个写好的东西暴露给Python，在集群上的main函数也用python写。因为现在的main函数包含的逻辑都很适合Python，例如:

* 如何读数据
* 训练多长时间，然后输出模型
* 如何输出log

之类的。

进而，如果我们的main函数在python里面写，那么用户在Python中按照我们定义的接口实现一个数据解析类(也就是用户实现的DataProvider)，然后再自定义一下数据放在哪里(例如HDFS)。那其实也不太需要C++里面的DataProvider。这么做好处有如下几个方面:

* Python的对象基本上都是可以序列化的。所以，直接将『数据解析类』序列化，发送给Worker就好了。
* 训练数据我们写一个HDFS文件列表的shuffle，切分功能。直接将每个worker应该训练的数据文件路径，也序列化发送给Worker就好了。
* 用户在集群上也可以和本地一样，完全用python驱动训练。这会增大用户使用的灵活性，减少单机使用和集群使用的差别。
* 去掉Paddle的DataProvider概念。

**Comment by wangkuiyi** [comment]
@reyoung 你说的是一种可行的思路。具体应该怎么做，我们还可以再想一想。

I think our discussion here has touched a critical design decision to be made -- should we encourage users to write readers for all kinds of their data, or should we make PP read a few supported types of data?

This question reminds me an experience with Google MapReduce and Hadoop MapReduce.  My understanding starts from Google MapReduce in my first job. Google MapReduce can read only three kinds of file formats -- text, SSTable, and RecordIO. (They added support to Bigtable later; but that's another story.) All these formats are "versatile containers", which means a file contains a big bundle of data instances. For example, we can do base64 encoding to audio clips and save them into a text file. Similarly, we can save video clips and images into protobuf records in SSTable and RecordIO files.  These few formats were proven enough for all MapReduce applications in Google.

After I left Google, I got to know Hadoop MapReduce, which encourages users to write their own readers by deriving a Java class from a Reader base class. This made the system very awkward -- tons of lines of code were introduced to support reading any kinds of data.  Moreover, the system efficiency is terribly bad -- many users reads literally their data saved as a bunch of small files -- without grouping small pieces of data into a big file, the reader causes the disk driver to seek and re-seek all the time, and the job performance downgrades because it reads slowly.

I think we are facing exactly the same problem with deep learning. When I was working on Deep Speech at AI Lab, we had to pack many short audio clips into a big file so to improve the reading performance; otherwise, we cannot read fast enough and feed the hungry GPUs.

If it is anyway right to pack data into big container files, PaddlePaddle might need to select a container to support other than allowing users to write their own readers or data providers.

If we decide to go this way, I would suggest that we create our own file format using protobuf format. An easy format is a sequence of {record_size, protobuf_record} pairs.

Also, we might want to define how to connect protobuf records to our data layers.   For example, to train an image classifier, we need images and labels:

```
message Instance {
  bytes[] image;
  int label;
}
```

Then we could define our data layer as:

```
image = paddle.layer.data(proto_field="image");
label = paddle.layer.data(proto_filed=label);
```

**Comment by reyoung** [comment]
> should we encourage users to write readers for all kinds of data, or should we make PP supports reading only a few file formats?

This question should be discussed widely. My opinion is we should let the user customize whole feeding data process, and there are several reasons.

* [Data Augumentation](https://www.kaggle.com/c/datasciencebowl/discussion/11360). Before neural network process input data, we often add some noises or randomization. If we make data in a static file, Paddle must implement all data augmentation methods, which might be a tremendous work.

* Neural network's input could be very complex and dynamic. In some research, a neural network's input could be even another neural network's output. To serialize them into a protobuf or another file format is time-consuming and not very straight forward.

So in my opinion,

* We could let user give a standard format of data as neural network's input, but we should let user customize the HOW to give them.
   * The standard format could be numpy ndarray. It will let the user write a Python function and return a numpy ndarray.
   * The standard format also could be protobuf(or I think HDF5 is better). But we must let users customize how to generate them, and give both batch and streaming ways to use them.
       * all static data could be stored in files, like Hadoop's map-reduce.
       * we also should provide a streaming way.

So it seems that whether to use protobuf or not is not the problem. The problem is `how to let user train neural network using streaming data or dynamic data`.

**Comment by reyoung** [comment]
我在想，这个功能可不可以用类似于Linux管道的方式实现？不过，和Linux管道有区别的地方是，这个管道可以是多机的，也是强类型的(protobuf)

Paddle只支持一种类型的输入，就是Protobuf。这个Protobuf的内容就是一种矩阵表示(也就是现在data能够接收的东西)，用户还是需要先把数据Encode成protobuf。

image

如果多机的话，这种管道可以交由其他框架搞，(例如redis或者kafka？这个我不太熟)，如果是单机的话，也可以是启动多进程，中间过一下单机的IPC之类的东西。

如果需要Data Augmentation或者其他操作，那么就在这个pipeline里面加入其他**进程**，输入是Paddle规定的Protobuf格式，输出也是这个格式。

image

因为，每一次encode或者augumentation，输出必须是Paddle的Protobuf类型，且是独立的进程，那么我们不再需要序列化什么函数，这种trick的方法。进程的移动，其实就用docker就好了。


这么做的好处是，逻辑简单清晰，也似乎满足了目前的需求，和未来online learning的需求

这么做的坏处是，

* 单机会变得略微麻烦。因为要启动多进程encode成protobuf。不过，这玩意应该也可以在python那里封装好了，用户感知不到？
* 做一次data augumentation就序列化反序列化一次，类似于装箱拆箱操作，并且还有多进程通信的开销，应该比较有开销。
   * 通信开销可能可以被计算隐藏掉
   * 可以不用protobuf？比如msgpack或者FlatBuffers?
       * msgpack 是个弱类型，没有类型信息
       * FlatBuffers因为整体是inplace的序列化，有个特别反人类的接口，用起来特别恶心。

**Comment by helinwang** [comment]
I want to add my thoughts on this topic. First I will state the topic I am discussing (maybe a sub-topic of this thread's topic):

How to make training on cluster requires minimal code change from training on local machine.

My proposal is on local machine we use:
```
data_provider = data_provide.imagenet(path="/local/", augment=my_augment_module)
optimizer.train(data_provider) 
```
to train on cloud we change the code to:
```
data_provider = data_provide.imagenet(
    path="data-path-or-name-on-cloud",
    augment=my_augment_module)
optimizer.train(
    data_provider,
    remote="paddle-cluster.com",
    env="data-augment-env")
```
In the above cloud training example, client program will connect to remote training cluster, send the model topology to server, also tell what data provider and augment module to use.

The problem is maybe the data provider and augment module is not available in remote cluster. That is when env parameter come into play. Implementation wise, it can be name of a docker image that is already pushed to docker image registry.

In this way, not everyone in team need to understand docker, just one person need to configure the correct environment (which contains the necessary data provider module and augment module). Other people can just need to use it.

And there could be already some data provider pre-defined in the default env (paddle official docker image).

This design uses the concept of environment as a paddle runtime. Environment is *program*. Model configuration (topology, data source name, how to augment, ...) is *data*. Different from lisp "program is data", we are using python / c++ which have a clear separation between program and data, I think we need to reflect this separation in our architecture and how we use our system.

**Comment by reyoung** [comment]
这里我们有了一点点共识。

与其序列化Python函数，不如直接用docker。将数据提供/data augumentation放到另一个进程里。

所以，数据提供还是要过多机器进程间的IPC？(也就是过网络？)

**Comment by helinwang** [comment]
@reyoung 我觉得我们可以提供一些基础的dataprovider，数据提供我想象的是，dataprovider的参数是数据集的名字(mnist, imagenet)，或者路径。所以感觉大多情况是走distributed file system而不是IPC。

**Comment by wangkuiyi** [comment]
In our new API design https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/api.md, we will not face this problem and don't need this solution.

**Pull request closed without merging**